### PR TITLE
Drop support for Python 3.7

### PR DESCRIPTION
**Pull request by Kludex**
Python 3.7 EOL is coming: https://endoflife.date/python.

### CODE REVIEW

**Review by adriangb — APPROVED** [review]

File: starlette/applications.py [hunk 1]
```diff
@@ -111,8 +111,7 @@ def build_middleware_stack(self) -> ASGIApp:
     def routes(self) -> typing.List[BaseRoute]:
         return self.router.routes
 
-    # TODO: Make `__name` a positional-only argument when we drop Python 3.7 support.
-    def url_path_for(self, __name: str, **path_params: typing.Any) -> URLPath:
+    def url_path_for(self, /, __name: str, **path_params: typing.Any) -> URLPath:
```
Comment (adriangb) [inline starlette/applications.py:114]:
```suggestion
    def url_path_for(self, /, name: str, **path_params: typing.Any) -> URLPath:
```

File: starlette/requests.py [hunk 2]
```diff
@@ -173,7 +173,7 @@ def state(self) -> State:
             self._state = State(self.scope["state"])
         return self._state
 
-    def url_for(self, __name: str, **path_params: typing.Any) -> URL:
+    def url_for(self, /, __name: str, **path_params: typing.Any) -> URL:
```
Comment (adriangb) [inline starlette/requests.py:176]:
```suggestion
    def url_for(self, /, name: str, **path_params: typing.Any) -> URL:
```

**Review by alex-oleshkevich — APPROVED** [review]

File: starlette/routing.py [hunk 3]
```diff
@@ -258,7 +258,7 @@ def matches(self, scope: Scope) -> typing.Tuple[Match, Scope]:
                     return Match.FULL, child_scope
         return Match.NONE, {}
 
-    def url_path_for(self, __name: str, **path_params: typing.Any) -> URLPath:
+    def url_path_for(self, /, __name: str, **path_params: typing.Any) -> URLPath:
```
Comment (adriangb) [inline starlette/routing.py]:
```suggestion
    def url_path_for(self, name: str, /, **path_params: typing.Any) -> URLPath:
```

Comment (adriangb) [reply starlette/routing.py]: Same for other occurrences

**Review by rkazinets — APPROVED** [review]

**Comment by Kludex** [comment]
I'll make the next release with 3.7, then I'll merge this.

File: starlette/applications.py [hunk 4]
```diff
@@ -111,9 +111,8 @@ def build_middleware_stack(self) -> ASGIApp:
     def routes(self) -> typing.List[BaseRoute]:
         return self.router.routes
 
-    # TODO: Make `__name` a positional-only argument when we drop Python 3.7 support.
-    def url_path_for(self, __name: str, **path_params: typing.Any) -> URLPath:
-        return self.router.url_path_for(__name, **path_params)
+    def url_path_for(self, /, name: str, **path_params: typing.Any) -> URLPath:
```
Comment (abersheeran) [inline starlette/applications.py]: Is there a mistake here?

Comment (Kludex) [reply starlette/applications.py]: It isn't supposed to be... Why do you think it is?

Comment (abersheeran) [reply starlette/applications.py]: I recall from our earlier discussion that `name` should be a positional-only parameter.

Maybe it should be changed to `(self, name: str, /, **path_params: typing.Any)`

Comment (Kludex) [reply starlette/applications.py]: Ah, it's a mistake! 🤣 

Thanks! 🙏

**Review by abersheeran — APPROVED** [review]

**Review by zanieb — APPROVED** [review]

**Comment by adriangb** [comment]
LGTM

**Pull request merged**